File: frontend/lib/services/auth_service.dart
import '../models/user.dart';
import '../services/api_service.dart';
import '../services/storage_service.dart';

class AuthService {
  final ApiService _api;
  AuthService(this._api);

  Future<Map<String, dynamic>> login(String email, String password) async {
    final data = await _api.post('/login', data: {
      'email': email.trim(),
      'password': password,
    });
    final token = data['token'] as String;
    final user = User.fromJson(data['user'] as Map<String, dynamic>);
    await StorageService.saveToken(token);
    await StorageService.saveUser(user);
    return {'token': token, 'user': user};
  }

  Future<void> logout() async {
    try {
      await _api.post('/logout');
    } catch (_) {
      // Even if the server call fails, clear local storage
    } finally {
      await StorageService.clearAll();
    }
  }

  Future<void> forgotPassword(String email) async {
    await _api.post('/forgot-password', data: {
      'email': email.trim(),
    });
  }

  Future<void> resetPassword({
    required String email,
    required String code,
    required String password,
    required String passwordConfirmation,
  }) async {
    await _api.post('/reset-password', data: {
      'email': email.trim(),
      'code': code.trim(),
      'password': password,
      'password_confirmation': passwordConfirmation,
    });
  }

  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
    required String newPasswordConfirmation,
  }) async {
    await _api.post('/change-password', data: {
      'current_password': currentPassword,
      'new_password': newPassword,
      'new_password_confirmation': newPasswordConfirmation,
    });
  }

  Future<User?> getStoredUser() async {
    return StorageService.getUser();
  }

  Future<String?> getStoredToken() async {
    return StorageService.getToken();
  }
}

File: ios/Runner/Services/AuthService.swift
import Foundation

struct AuthService {
    private let api = APIService.shared

    // MARK: - Login

    struct LoginResponse: Decodable {
        let token: String
        let user: User
    }

    func login(email: String, password: String) async throws -> (token: String, user: User) {
        let resp: LoginResponse = try await api.post("/login", body: [
            "email":    email.trimmingCharacters(in: .whitespaces),
            "password": password
        ])
        try KeychainService.save(resp.token, for: AppConfig.tokenKey)
        if let userData = try? JSONEncoder().encode(resp.user) {
            UserDefaults.standard.set(userData, forKey: AppConfig.userKey)
        }
        return (resp.token, resp.user)
    }

    // MARK: - Logout

    func logout() async {
        try? await api.postVoid("/logout")
        KeychainService.delete(for: AppConfig.tokenKey)
        UserDefaults.standard.removeObject(forKey: AppConfig.userKey)
    }

    // MARK: - Password flows

    func forgotPassword(email: String) async throws {
        try await api.postVoid("/forgot-password", body: [
            "email": email.trimmingCharacters(in: .whitespaces)
        ])
    }

    func resetPassword(email: String, code: String,
                       password: String, passwordConfirmation: String) async throws {
        try await api.postVoid("/reset-password", body: [
            "email":                 email.trimmingCharacters(in: .whitespaces),
            "code":                  code.trimmingCharacters(in: .whitespaces),
            "password":              password,
            "password_confirmation": passwordConfirmation
        ])
    }

    func changePassword(currentPassword: String,
                        newPassword: String,
                        newPasswordConfirmation: String) async throws {
        try await api.postVoid("/change-password", body: [
            "current_password":          currentPassword,
            "new_password":              newPassword,
            "new_password_confirmation": newPasswordConfirmation
        ])
    }

    // MARK: - Local storage helpers

    func storedUser() -> User? {
        guard let data = UserDefaults.standard.data(forKey: AppConfig.userKey) else { return nil }
        return try? JSONDecoder().decode(User.self, from: data)
    }

    func storedToken() -> String? {
        try? KeychainService.read(for: AppConfig.tokenKey)
    }

    func clearStorage() {
        KeychainService.delete(for: AppConfig.tokenKey)
        UserDefaults.standard.removeObject(forKey: AppConfig.userKey)
    }
}
